### PR TITLE
Abstract vision framework with included support for focus, navigator object and browse mode caret highlighting

### PR DESCRIPTION
**Pull request by LeonarddeR**
Pr provided on behalf of @babbagecom

### Link to issue number
Implements #971, but without a gui

### Background
For years, NVDA has been an open source screen reader for the blind and partially sighted people who rely on braille and speech output. There are no built-in facilities within NVDA to manipulate the contents of the screen. There are several outstandign issues asking for such features, such as:

* #7857: Making the screen black (i.e. a screen curtain) while NVDA is active, mainly for privacy reasons
* #971: Visual highlight of focus, review or browse mode caret location

There is also a long standing desire to have a basic screen magnification facility within NVDA.

This pr intends to lay the base of a vision framework that can be used to implement such functionality in the core of NVDA. Though there is no GUI in the current pull request, the framework is fully working. The GUI will be added in a follow up pr.

### A new vision module
Most magic is supposed to happen within a new vision module. The vision module consists of two distinct concepts:

* Vision enhancement providers: providers of functionality that change the contents of the screen. A vision enhancement provider can be a magnifier, highlighter, or whatever else that changes the screen contents.

	This pr contains a highlighter based on a transparent overlay window on which lines are drawn around the several objects on screen.

	The [Screen Curtain add-on](https://addons.nvda-project.org/addons/screenCurtain.en.html), which is a prototype add-on written by me on behalf of @babbagecom and ported to an add-on by @josephsl, is supposed to be added to NVDA in subsequent pr.

	Vision enhancement providers will also be able to benefit from the functionality to set driver/provider specific options in the gui (#8214).

* The vision handler, which receives information about focus, foreground, navigator object, review and mouse changes from the core of NVDA. It keeps track of the location of objects on screen and communicates this information to one or more active vision enhancement providers. This information is communicated to vision enhancement providers using extension points. The handler keeps track of active enhancement providers and can activate and deactivate them. Therefore, it is very much similar to the braille handler in its approach.

### Follow up strategy
@feerrenrut has requested for this framework to be filed into subsequent pull requests. After this pr has been fully reviewed and merged, I intent to file at least two additional pr's:

1. A graphical interface for the framework
2. A built-in screen curtain

### Known issues
The current highlighter colors might be difficult for some people to see. In the future, I'd like this to be configurable.

### Changelog entries:
* Changes for developers
    + A vision framework has been added to NVDA. It allows you to create vision enhancement providers; modules that can change screen contents, optionally based on input from NVDA about object locations.
        * Add-ons can bundle their own providers in a visionEnhancementProviders folder.
        * See the vision and visionEnhancementProviders modules for the implementation of the framework and examples, respectively.

### CODE REVIEW

**Comment by Adriani90** [comment]
Support from @nishimotz is very welcome I think. As part of the japanese NVDA Team, he is the creator of the add-on focus highlight
https://github.com/nvdajp/focusHighlight
I am sure that the japanese NVDA Team can bring much knowledge in this work as well.

**Comment by nishimotz** [comment]
I am testing the AppVeyor build (nvda_snapshot_pr9064-16441,7aaee3c0.exe)
and confirmed the default highlighter works if nvda.ini is modified something like this

```
[vision]
        highlighter = NVDAHighlighter
        [[NVDAHighlighter]]
                highlightFocus = true
                highlightCaret = true
                highlightNavigatorObj = true
```

File: source/vision/__init__.py
Comment (nishimotz) [inline source/vision/__init__.py]: I got errors as follows regarding this line:
```
ERROR - queueHandler.flushQueue (13:47:02.601):
Error in func _loadBufferDone from eventQueue
Traceback (most recent call last):
  File "queueHandler.pyc", line 53, in flushQueue
  File "virtualBuffers\__init__.pyc", line 464, in _loadBufferDone
  File "browseMode.pyc", line 1152, in event_treeInterceptor_gainFocus
  File "browseMode.pyc", line 1465, in event_gainFocus
  File "vision\__init__.pyc", line 601, in handleGainFocus
  File "vision\NVDAHighlighter.pyc", line 63, in updateContextRect
  File "vision\__init__.pyc", line 270, in updateContextRect
  File "vision\__init__.pyc", line 190, in getContextRect
  File "vision\__init__.pyc", line 205, in _getRectFromTextInfo
NameError: global name 'self' is not defined
```

**Comment by nishimotz** [comment]
nvda_snapshot_pr9064-16574,1ca723ff.exe fails initialization as follows:

```
INFO - core.main (20:56:31.969):
NVDA version pr9064-16574,1ca723ff
INFO - core.main (20:56:31.969):
Using Windows version 10.0.17134 workstation

INFO - core.main (20:56:33.881):
NVDA initialized
ERROR - unhandled exception (20:56:33.881):
Traceback (most recent call last):
  File "wx\core.pyc", line 3240, in <lambda>
  File "vision.pyc", line 545, in setProvider
  File "vision.pyc", line 440, in getProvider
ImportError: No module named NVDAHighlighter
```

I cannot import this via Python console as well:

```
>>> import visionEnhancementProviders.NVDAHighlighter
Traceback (most recent call last):
  File "<console>", line 1, in <module>
ImportError: No module named NVDAHighlighter
```

**Comment by leonardder** [comment]
Thanks @nishimotz! This should have been fixed in the most recent code.

**Comment by leonardder** [comment]
While the highlighter based on WX is doing its job pretty well in daily use, a major eyesore is that it's being hosted in NVDA's main thread. I've seen cases where the system freezes entirely when the log viewer is opened with the highlighter on.

I guess the only way to avoid this, is by using a separate thread for the highlight, like Focus Highlight also does. However, I guess that also means we have to abandon wx for this. Thoughts @nishimotz?

**Comment by nishimotz** [comment]
I'm sorry to know that the highlighter working in the main thread is difficult.
Is there no option to use WX for drawing in a separate thread?

Another concern is that the current version of Focus Highlight uses GDI Plus because it seems easier to handle multiple monitors at physical coordinates.

**Comment by leonardder** [comment]
> Is there no option to use WX for drawing in a separate thread?

I'm afraid we can't even create a window on a separate thread, but I'll do some further investigation.

> Another concern is that the current version of Focus Highlight uses GDI Plus because it seems easier to handle multiple monitors at physical coordinates.

Are you saying this is a concern for focus highlight, or for the wx based highlighter? Actually, apart from focus highlight not written very object orientedly, I believe it does it job quite well.

**Comment by leonardder** [comment]
nishimotz: I updated the highlighter to use GDI Plus, based on your work but a bit more pythonic. I wonder what you think about the current approach. Also, tests would be greatly appreciated.

**Comment by nishimotz** [comment]
Found an issue with nvda_snapshot_pr9064-17053,91896a0a.exe.

steps:

* create nvda snapshot as portable copy
* modify userConfig/nvda.ini for enabling vision framework
* start nvda
* start Chrome and visit a page, move focus to a link, for example
* press NVDA+space twice
* an error sound

```
ERROR - scriptHandler.executeScript (14:36:13.503):
error executing script: <bound method GlobalCommands.script_toggleVirtualBufferPassThrough of <globalCommands.GlobalCommands object at 0x04CB77D0>> with gesture u'NVDA+\u30b9\u30da\u30fc\u30b9'
Traceback (most recent call last):
  File "scriptHandler.pyc", line 196, in executeScript
  File "globalCommands.pyc", line 1261, in script_toggleVirtualBufferPassThrough
  File "treeInterceptorHandler.pyc", line 141, in _set_passThrough
  File "vision.pyc", line 630, in handleGainFocus
  File "visionEnhancementProviders\NVDAHighlighter.pyc", line 107, in updateContextRect
  File "vision.pyc", line 267, in updateContextRect
  File "vision.pyc", line 188, in getContextRect
AttributeError: 'ChromeVBuf' object has no attribute 'location'
```

So far, the highlight works nicely, as well as the reviewing character position.

**Comment by leonardder** [comment]
Thanks for reporting this. This should now be fixed.

How do you perceive the performance of this highlighter in comparison 
with early code in this pr and your own focus highlight add-on?

**Comment by dingpengyu** [comment]
hi leonardder
Can you fix Failure: AppVeyor was unable to build non-mergeable pull request?

**Comment by leonardder** [comment]
Thanks for pointing this out!

**Comment by nishimotz** [comment]
Confirmed that the Chrome passThrough issue is addressed.

I have found the difference between the results of vision framework and that of my add-on.

tested version: nvda_snapshot_pr9064-17078,432ec1ef.exe

steps:

* launch NVDA
* NVDA+N to open the menu
* press C to open the configuration profiles dialog
* press NVDA+B and wait until the speech is finished
* press NVDA+Shift+O (laptop) or NVDA+numpad 5 (desktop)
* it can be confirmed that the navigator object is the close button (right bottom)

The focus highlight add-on can show:
* focus is on the list item
* navigator object is the close button

Vision framework can show:
* only the focus, which is on the list item

**Comment by leonardder** [comment]
I haven't looked at the code yet. However, this might be because focus 
highlight add-on continuously polls the navigator object, while the 
vision framework is event driven, i.e. if handleReviewMove isn't called 
explicitly, the vision framework doesn't know about the new navigator 
object.


Actually, I'm not sure whether this is a bug. The script help for NVDA+b 
clearly states that it speaks the window. I don't think the new 
navigator object is reflected in braille either, though I haven't yet 
tested this. Definitely something I should look into.

**Comment by dingpengyu** [comment]
hello leonardder
I noticed that this pr is not categorized in the AfterPython3Transition tag. Does this indicate that it might be merged into the master? Also, what problem is there still not solved in this pr?

**Comment by nishimotz** [comment]
nvda_snapshot_pr9064-17306,32d9232f.exe raises errors as follows.

Should I modify nvda.ini?

nvda.log

```
INFO - __main__ (12:38:48.187):
Starting NVDA
INFO - core.main (12:38:49.013):
Config dir: C:\Users\nishimotz\AppData\Roaming\nvda
INFO - config.ConfigManager._loadConfig (12:38:49.013):
Loading config: C:\Users\nishimotz\AppData\Roaming\nvda\nvda.ini
INFO - core.main (12:38:49.030):
Developer Scratchpad mode enabled
INFO - core.main (12:38:49.086):
NVDA version pr9064-17306,32d9232f


ERROR - windowUtils._rawWindowProc (12:38:50.180):
Error in wndProc
Traceback (most recent call last):
  File "windowUtils.pyc", line 253, in _rawWindowProc
  File "visionEnhancementProviders\NVDAHighlighter.pyc", line 188, in windowProc
  File "visionEnhancementProviders\NVDAHighlighter.pyc", line 234, in _paint
AttributeError: 'HighlightWindow' object has no attribute 'location'


ERROR - eventHandler.executeEvent (12:38:50.513):
error executing event: gainFocus on <NVDAObjects.Dynamic_IAccessibleRichEdit50WindowNVDAObject object at 0x049E16B0> with extra args of {}
Traceback (most recent call last):
  File "eventHandler.pyc", line 155, in executeEvent
  File "eventHandler.pyc", line 92, in __init__
  File "eventHandler.pyc", line 100, in next
  File "appModules\nvda.pyc", line 86, in event_gainFocus
  File "eventHandler.pyc", line 100, in next
  File "NVDAObjects\behaviors.pyc", line 179, in event_gainFocus
  File "NVDAObjects\__init__.pyc", line 1036, in event_gainFocus
  File "vision\visionHandler.pyc", line 206, in handleGainFocus
NameError: global name 'mightHaveCaret' is not defined
```


vision group of nvda.ini

```
[vision]
	magnifier = None
	highlighter = NVDAHighlighter
	colorEnhancer = None
	[[NVDAHighlighter]]
		highlightFocus = True
		highlightCaret = True
		highlightNavigatorObj = True
```

**Comment by feerrenrut** [comment]
This re-organisation looks a lot better to me. I think it would be worth adding a readme.md to the vision package. This can be helpful to explaing the layout a little, but also the concepts. I'm visualizing help and examples for add-on developers eventually as well. Perhaps this should be in the docscring for the package. Though for indepth information this could make the file mostly comments. Happy to hear opinions on where this should live.

**Comment by leonardder** [comment]
There are more github projects using per directory readme.md files, yet 
it makes inspection of the package from the python console a whole lot 
more difficult, as the information is not available using that method.

**Comment by nishimotz** [comment]
nvda_snapshot_pr9064-17312,c3faf25e.exe still raises errors as follows.

```
ERROR - eventHandler.executeEvent (11:22:22.559):
error executing event: gainFocus on <NVDAObjects.Dynamic_IAccessibleRichEdit50WindowNVDAObject object at 0x04BEE530> with extra args of {}
Traceback (most recent call last):
  File "eventHandler.pyc", line 173, in executeEvent
  File "eventHandler.pyc", line 110, in __init__
  File "eventHandler.pyc", line 118, in next
  File "appModules\nvda.pyc", line 86, in event_gainFocus
  File "eventHandler.pyc", line 118, in next
  File "NVDAObjects\behaviors.pyc", line 179, in event_gainFocus
  File "NVDAObjects\__init__.pyc", line 1036, in event_gainFocus
  File "vision\visionHandler.pyc", line 206, in handleGainFocus
NameError: global name 'mightHaveCaret' is not defined
```

**Comment by nishimotz** [comment]
Tested nvda_snapshot_pr9064-17320,48bc1b0b.exe.
Very stable and reliable.

When Google Chrome opens a website, switch to browse mode, and press H to go to the first heading, the navigator object is shown as the first character of the heading element.

Is this intentional? 

Vision framework:

image

Focus Highlight:

image

**Comment by leonardder** [comment]
If everything is working as expected, you should see a box around the 
navigator object, and a box around the first character of the heading 
that highlights the currently selected character in browse mode. The 
latter box should change its position when you move with the left and 
right arrow keys from character to character.

**Comment by k-kolev1985** [comment]
Sorry for the maybe stupid question, but is there a way for me to test this? Is it implemented as a hidden feature in current "alpha" snapshots?

**Comment by leonardder** [comment]
@feerrenrut I think I have applied all the changes that we discussed by voice earlier. Could you have another look?

File: source/vision/visionHandler.py
Comment (feerrenrut) [inline source/vision/visionHandler.py]: Could `post_configProfileSwitch` be triggered before the `wx.CallAfter`, short-circuiting this logic. Perhaps instead:

```
		wx.CallAfter(self.postGuiInit)

	def postGuiInit(self):
		self.handleConfigProfileSwitch
		config.post_configProfileSwitch.register(self.handleConfigProfileSwitch)
```

File: source/vision/visionHandler.py
Comment (feerrenrut) [inline source/vision/visionHandler.py]: Beware, this leaves the `providers` list missing `providerInstance`, but returns indicating that it can't be removed. The other `return False` has the same issue. This may be intentional, in which case I think it deserves a comment about why it's what we want.

File: source/vision/visionHandler.py
Comment (feerrenrut) [inline source/vision/visionHandler.py]: What does the return mean? I assume successfully terminated, what does an unsuccessful termination mean?

File: source/vision/visionHandler.py
Comment (feerrenrut) [inline source/vision/visionHandler.py]: This is a handy trick.

File: source/vision/visionHandler.py
Comment (feerrenrut) [inline source/vision/visionHandler.py]: I think this deserves an explanation of "why": "During termination of a provider, the instance is not destroyed (for efficiency in slow to start providers), we cant rely on providers to de-register themselves, so instead we recreate the extensionPoints and ask active providers to re-register."

File: source/vision/providerBase.py
Comment (feerrenrut) [inline source/vision/providerBase.py]: I would expand this comment to mention that it should only register with the extension points, it may be called at several times between initialisation and termination.

File: source/vision/visionHandlerExtensionPoints.py
Comment (feerrenrut) [inline source/vision/visionHandlerExtensionPoints.py]: Some docs here, especially to document the parameters passed in.

File: source/visionEnhancementProviders/NVDAHighlighter.py
Comment (feerrenrut) [inline source/visionEnhancementProviders/NVDAHighlighter.py]: Please specify the functions used, or import as something.

File: source/visionEnhancementProviders/NVDAHighlighter.py
Comment (feerrenrut) [inline source/visionEnhancementProviders/NVDAHighlighter.py]: these aren't used

File: source/visionEnhancementProviders/NVDAHighlighter.py
Comment (feerrenrut) [inline source/visionEnhancementProviders/NVDAHighlighter.py]: not used

File: source/visionEnhancementProviders/NVDAHighlighter.py
Comment (feerrenrut) [inline source/visionEnhancementProviders/NVDAHighlighter.py]:
```suggestion
		A higher width reduces the inner dimensions of the rectangle.
```

File: source/visionEnhancementProviders/NVDAHighlighter.py
Comment (feerrenrut) [inline source/visionEnhancementProviders/NVDAHighlighter.py]:
```suggestion
		A higher margin increases both the inner and outer dimensions of the highlight's rectangle.
```

File: source/visionEnhancementProviders/NVDAHighlighter.py
Comment (feerrenrut) [inline source/visionEnhancementProviders/NVDAHighlighter.py]:
```suggestion
		Therefore, if you need to increase the outer dimensions of the rectangle, you need to increase the margin as well.
```

File: source/visionEnhancementProviders/NVDAHighlighter.py
Comment (feerrenrut) [inline source/visionEnhancementProviders/NVDAHighlighter.py]: Please pull these colours and styles out into constants.

EG:
```
BLUE = RGB(0x03, 0x36, 0xff)
PINK = RGB(0xff, 0x02, 0x66)
YELLOW = RGB(0xff, 0xde, 0x03)

DASH_BLUE =  HighlightStyle(BLUE, 5, winGDI.DashStyleDash, 5)
SOLID_PINK = HighlightStyle(PINK, 5, winGDI.DashStyleSolid, 5)
SOLID_BLUE = HighlightStyle(BLUE, 5, winGDI.DashStyleSolid, 5)
SOLID_YELLOW = HighlightStyle(YELLOW, 2, winGDI.DashStyleSolid, 2)
```

File: source/visionEnhancementProviders/NVDAHighlighter.py
Comment (feerrenrut) [inline source/visionEnhancementProviders/NVDAHighlighter.py]: Whats this for?

File: source/vision/visionHandler.py
Comment (feerrenrut) [inline source/vision/visionHandler.py]: I think `__import__` expects a list for this arg, but gets a tuple.

**Review by feerrenrut — CHANGES_REQUESTED** [review]
This is looking good to me, lots of small comments, but I expect this will be good to go after this.

File: source/scriptHandler.py
Comment (leonardder) [inline source/scriptHandler.py]: I can't find your initial comment about this, but I revisited this. Now, global plugins and appModules take precedence over vision enhancement provider scripts. I think this makes more sense than it was before.

File: source/vision/__init__.py
Comment (leonardder) [inline source/vision/__init__.py]: You had a comment about using the word check as a name for the function. I'm very reluctant to change this as in fact, check is a method on driverHandler.Driver. In other words, check is used this way for BrailleDisplayDrivers and SynthDrivers as well. I think I prefer consistency over clarity here.

File: source/vision/util.py
Comment (leonardder) [inline source/vision/util.py]: This was getReviewRectangle before. I decided to rename these in order to improve consistency.

**Comment by leonardder** [comment]
@feerrenrut: Thanks for your extensive review! I think I've addressed all the comment, particularly by adding many comments and additional doc strings.

**Comment by josephsl** [comment]
Hi, one small favor: can you test if this is Python 3 compliant for future-proofing purposes? Thanks.

**Comment by leonardder** [comment]
I've tried to ensure compatibility, but as I proposed elsewhere, I think 
we should do the merge of this into threshold by hand, preferably after 
threshold_py3_staging went into threshold.

File: source/vision/__init__.py
Comment (feerrenrut) [reply source/vision/__init__.py]: Consistently unclear? I missed that it's on the `Driver` base class.

This is an opportunity to adapt the interface, and provide further clarity. Consider the following psudo-code approach:

``` Python
class VisionEnhancementProvider(driverHandler.driver):

	@abstractclassmethod
	def canStart():
		raise NotImplementedError

	def check():
		return self.canStart()

```
You might even want to re-map `check` to `canStart` in `__new__` or `__init__` if you wanted to save the extra function call overhead, but I doubt the performance cost justifies the cost to readability.

Comment (leonardder) [reply source/vision/__init__.py]: Unfortunately, there is no such thing as an abstract class method.

HOw about implementing canStart for braille and speech as well?

Then:

* We rename check to canStart
* Calling check  raises a deprecation warning
* The base method of canStart calls check on classes where check is implemented/available.

File: source/vision/visionHandlerExtensionPoints.py
Comment (feerrenrut) [inline source/vision/visionHandlerExtensionPoints.py]: minor typo: tehy -> they

Comment (feerrenrut) [reply source/vision/visionHandlerExtensionPoints.py]: This typo looks like it has been copied to a few of these comments, a search and replace should fix this.

File: source/vision/constants.py
Comment (feerrenrut) [inline source/vision/constants.py]: minor typo: trigggered -> triggered

File: source/vision/__init__.py
Comment (leonardder) [reply source/vision/__init__.py]: May be isAvailable is a better name then canStart? see the doc string of check on driverHandler.Driver

**Comment by feerrenrut** [comment]
> we should do the merge of this into threshold by hand, preferably after threshold_py3_staging went into threshold.

Yes, this makes sense. I think it's best to merge the py 3 changes into this when you are ready, commit that even with errors, then commit fixes for any errors on top. You can wait for threshold to have py 3 or merge threshold_py3_staging, I won't squash merge to threshold, I think the history will be useful

**Comment by leonardder** [comment]
I will wait for threshold to have threshold_py3_staging. It might even 
be best to wait until master is py3, as otherwise we yet might get merge 
conflicts.

File: source/vision/__init__.py
Comment (feerrenrut) [reply source/vision/__init__.py]: A [stackoverflow answer](https://stackoverflow.com/a/11218474/346188) presents a solution for an abstract class method. This won't stop you from calling 
`VisionEnhancementProvider.canStart()`, you need `raise NotImplementedError` for that, but it will stop you from creating an instance of the class that does not override the class method. It also communicates the intent quite well.

> HOw about implementing canStart for braille and speech as well?

I think I would prefer not to extend the scope of the PR that far, it would logically follow from there that we get rid of `check` entirely. 

> May be isAvailable is a better name then canStart? see the doc string of check on driverHandler.Driver

Perhaps, but only if this really fits what we want the providers to tell us, this method should be adapting the generic `Driver.check` function, to make more sense in the context of a `VisionEnhancementProvider`. I think we want them to tell us more than "are they available", but also if they can actually be started / run. Say two providers require the same resources, they can't both run at once, this is their chance to say they are not able to. Am I understanding the design correctly?

**Comment by leonardder** [comment]
@feerrenrut: I've tried up this branch a bit after the Python 3 merge. At least all relevant new commits since your last view of this pr are now one after the other.
Particularly python 3 changes have been applied.

**Review by feerrenrut — APPROVED** [review]
Ok looks good

**Comment by feerrenrut** [comment]
Except there is a failing unit test, "No module named 'wx'"

**Comment by leonardder** [comment]
Ah, I'm sorry for that. This all has to do with the linter complaining about module imports have to be at the top of the file.

**Comment by leonardder** [comment]
@feerrenrut I'm afraid I had to use some noqa in the unit module to get this work. This is because we have to change the path before importing modules, and this is something Flake8 doesn't agree with.

The sourceEnv noqa is intentional, as this module is only imported to manipulate the path.

File: tests/unit/__init__.py
Comment (feerrenrut) [inline tests/unit/__init__.py]: I'm fine with using `noqa` to suppress this case, however please include a comment about why and a hint about what the lint error is. It's not much fun for a future developer to have to look these up and understand why it is like this.

Comment (leonardder) [reply tests/unit/__init__.py]: What would be proper syntax for this? I assume something like this is not going to work

```suggestion
import sourceEnv  # noqa: F401  module imported but unused
```

**Comment by leonardder** [comment]
Mm, I was ignoring linter warning F402, which had to be E402. Interesting enough, Ignoring F402 also ignored E402 unexpectedly, that's why I didn't see this before.

File: tests/unit/__init__.py
Comment (feerrenrut) [reply tests/unit/__init__.py]: Yep something like that works, though it gets messy if you have multiple errors to ignore. 

I would do:
```
# F401: module imported but unused. We need sourceEnv for unit tests because of ?
import sourceEnv  # noqa: F401
```

Comment (leonardder) [reply tests/unit/__init__.py]: I think I've done something along those lines.

File: tests/unit/__init__.py
Comment (feerrenrut) [inline tests/unit/__init__.py]: From a unit testing perspective it would be better if we can avoid this. Which part of the vision framework relies on a wx app being available? Can we mock this part instead?

**Review by feerrenrut — APPROVED** [review]
Generally this looks good. But I am a little worried about the dependencies of vision, if possible I would prefer it rely on dependency injection if possible. Can we start by identifying what fails if wx is missing?

**Comment by leonardder** [comment]
wx is used to make sure that the gui is initialized (CallAfter) before initializing providers. This, because providers might rely on the gui. The highlit"ghter, for example, relies on gui.mainFrame.handle to be the parent of its window.
By default, there is no provider active. So it might make sense to check for configured providers before using wx.CallAfter. If we ever decide to write a pseudo provider, this requires another revisit.
We can also mock wx.CallAfter, e.g. 'lambda func, *args, **kwargs: func(*args, **kwargs)'. Other code might benefit from this too.

**Review by feerrenrut — APPROVED** [review]

**Pull request merged**
